File: javapractice/src/main/java/com/zafar/javapractice/designpattern/FacadeDpExample.java
package com.zafar.javapractice.designpattern;

/*
Facade pattern hides the complexities of the system and provides an interface to the client using which the client can access the system.
This type of design pattern comes under structural pattern as this pattern adds an interface to existing system to hide its complexities.
This pattern involves a single class which provides simplified methods required by client and delegates calls to methods of existing system classes.
*/

public class FacadeDpExample {
    public static void main(String args[]) {
        ShapeMaker shapeMaker = new ShapeMaker();
        shapeMaker.drawCircle();
        shapeMaker.drawRectangle();
        shapeMaker.drawSquare();
    }
}

interface Shape {
    void draw();
}

class Rectangle implements Shape {
    @Override
    public void draw() {
        System.out.println("drawing rectangle");
    }
}

class Square implements Shape {
    @Override
    public void draw() {
        System.out.println("drawing square");
    }
}

class Circle implements Shape {
    @Override
    public void draw() {
        System.out.println("drawing circle");
    }
}


class ShapeMaker {
    private Shape rectangle;
    private Shape circle;
    private Shape square;

    public ShapeMaker() {
        rectangle = new Rectangle();
        circle = new Circle();
        square = new Square();
    }

    public void drawCircle() {
        circle.draw();
    }

    public void drawRectangle() {
        rectangle.draw();
    }

    public void drawSquare() {
        square.draw();
    }
}

File: kotlinpractice/src/main/java/com/zafar/kotlinpractice/dp/FacadeDpExample.kt
package com.zafar.kotlinpractice.dp


fun main(args: Array<String>) {
 val shapeMaker = ShapeMaker()
    shapeMaker.drawCircle()
    shapeMaker.drawRectangle()
    shapeMaker.drawSquare()
}

interface Shape {
    fun draw()
}

class Rectangle : Shape {
    override fun draw() {
        println("drawing rectangle")
    }
}

class Square : Shape {
    override fun draw() {
        println("drawing square")
    }
}

class Circle : Shape {
    override fun draw() {
        println("drawing circle")
    }
}

class ShapeMaker {
    private var rectangle: Shape? = null
    private var circle: Shape? = null
    private var square: Shape? = null

   constructor() {
       rectangle = Rectangle()
       circle = Circle()
       square = Square()
   }

    fun drawCircle() {
        circle?.draw()
    }

    fun drawRectangle() {
        rectangle?.draw()
    }

    fun drawSquare() {
        square?.draw()
    }
}